### ISSUE TITLE
Would it be possible to split out "libmcli" into a separate project?

### ISSUE THREAD

**Issue by M-Reimer** [vdr-projects/vdr-plugin-mcli#13]
Haven't tried so far. Maybe it would also be possible to get what I want using "creative fiddling with build parameters". And if you prefer to not split off "libmcli", I'll probably do just that.

My "issue" is that the "libmcli" part can optionally also be used by "minisatip". Currently I'm building "libmcli" directly from a mirror of the Baycom files: https://github.com/VDR4Arch/vdr4arch/blob/master/deps/libmcli/PKGBUILD
And make minisatip depend on it.

Currently it's my "libmcli" package which delivers the "netceiver tools". So building vdr-plugin-mcli including all tools will conflict with my libmcli package.

As "libmcli" is at least used by two projects, I think the proper way of building would be to finally have 3 packages:

libmcli:
- libmcli shared object (.so) file
- pkg-config file so other projects can detect the library
- all the tools to interact and update the netceiver

vdr-plugin-mcli:
- Depends on libmcli
- Contains just the VDR plugin

minisatip:
- Depends on libmcli

What do you think about this?

**Comment by pbiering** [comment]
Generally a good idea to share code/libraries, but this also creates new dependencies.

The VDR plugin and the 3 created binaries have `libmcli` statically linked so far, e.g.

```
ldd ./libvdr-mcli.so |grep mcli
(empty)
```

`./mcast/client/libmcli.so` is created during compilation, but not used.

=> it would potentially require some rework to change the code from statically linked to dynamically linked incl. version check and separate the library itself.

BTW: where did you run into a "conflict"? On a package manager level it would be easy to create a subpackage (e.g. in RPM: e.g. vdr-mcli-tools) as at least a workaround....and also create a subpackage for the "libmcli.so" for to be used by others.

**Comment by M-Reimer** [comment]
> BTW: where did you run into a "conflict"? On a package manager level it would be easy to create a subpackage (e.g. in RPM: e.g. vdr-mcli-tools) as at least a workaround....and also create a subpackage for the "libmcli.so" for to be used by others.

This is totally OK if you see it from a distributor perspective. If you want the VDR plugin in your repo, then you build it anyway so you can split off the additional things in that process.

But for Arch noone really is a "distributor" in the "classic way". Meaning that there are no binary packages for VDR stuff. All that I offer are PKGBUILDs which are published to AUR for Arch Linux users.

And now it gets difficult. If everything, someone needs, is minisatip, then he has to build the whole VDR plugin first to get the split package. And to do that he has to build VDR, too. Even if he doesn't want VDR or the plugin at all.

Splitting off libmcli would probably also make it more common that distributions, shipping minisatip, also enable NetCeiver support. If everything, that's needed for that, is a small, simple to build, library, then this would be way easier.

How do you think about the netceiver tools? Own repo called "netceiver-tools" or place them into the "libmcli" repo? I think tools and library are that small that a distributor could easily split this off if he wants to. But in general I think it would be good practice to package the tools in their own package for the case that someone just gets the libmcli pulled in as dependency and doesn't actually need the netceiver support offered by it. In this case he would not need the tools.

If we can agree on some model, then I would put this on my TODO list.

**Comment by pbiering** [comment]
What about having (because "mcli" is already used somehow on GitHub) an ` netceiver` source package resulting then in RPM packaging in `netceiver-libs`, `netceiver-tools` and `netceiver-devel`?

`netceiver-libs` can be then used by `minisatip` and upcoming shrinked `vdr-mcli` (after carve-out of "mcli" code and switched to dynamically linked functions and requiring then for build process `netceiver-devel`) and all which requires the tools also can install the `netceiver-tools` package.

Then only one new repo is required with name`netceiver`.

**Comment by M-Reimer** [comment]
Where is "mcli" already used? Minor problem, I see, is that "libmcli" is already somewhat established.

**Comment by pbiering** [comment]
> Where is "mcli" already used?

Found by searching through "github" repositories.

> Minor problem, I see, is that "libmcli" is already somewhat established.

Is there any other server/unit beside "netceiver" which is serving video/audio streams in same manner? If not, `netceiver-client-libs`, `netceiver-client-devel` and `netceiver-client-tools` would be also an option.

**Comment by M-Reimer** [comment]
You are right. Only "netceiver" uses the mcli protocol and probably noone else will ever use it as "Sat>IP" is well established.

What about calling the whole library project "libnetceiver" then? The tools could go into that, too. The makefile could get different install targets to simplify splitting.

**Comment by pbiering** [comment]
> What about calling the whole library project "libnetceiver" then? The tools could go into that, too. The makefile could get different install targets to simplify splitting.

Your suggestion is also possible (found on Fedora some lib* packages having also lib*-tools|utils available)

```
yum list 'lib*' | grep tools  | awk -F . '{ print $1 }' | sort -u | wc -l
75

yum list 'lib*' | grep utils  | awk -F . '{ print $1 }' | sort -u | wc -l
56
```
but there are a lot more packages with schema like e.

- net-snmp-libs + net-snmp-utils

```
declare -A libs; libs=( $( yum list '*-libs' '*-tools' '*-utils' | awk -F . '{ print $1 }' | sort -u ) ); for p in ${!libs[@]}; do b=${p/-libs}; if [ -n ${libs[$p-utils]} ]; then echo "utils"; elif [ -n ${libs[$p-tools]} ]; then echo "tools"; fi; done | sort | uniq -c
   1200 utils
```

=> I would prefer the 2nd

repo name: netceiver-client (and name of SRPM) - this avoids also collision if the netceiver-server code will appear here suddenly
library package: netceiver-client-libs
tools package: netceiver-client-utils
devel  package: the header files, .a libraries and pkgconfig files if required

**Comment by M-Reimer** [comment]
Maybe "netceiver-libs" as primary repo then as technically the repo does not hold an actual standalone client (that would be minisatip or the VDR plugin)?

Problem is that it is basically the responsibility of the distributor to "split up" into packages and if a distributor does not want to split anything (Arch for example has no "devel" packages) then I would prefer having a "netceiver-libs" package as the "catch all" package over a "netceiver-client" package which does not even hold something that qualifies as "client" by itself.

**Comment by M-Reimer** [comment]
I guess I'll compare a few existing projects with similar splitting this weekend to see how they did it.

When splitting then better do it once and in a good way.

**Comment by M-Reimer** [comment]
I had a look at the possible candidates on Arch (maybe you can easily verify this for a Debian based distribution. I don't use any Debian based distributions here). On Arch packages with "-libs" are pretty rare actually. And in every single case, that I tested, the "-libs" package is an optional package which is shipped with some kind of "host package".

For example "bluez-libs" is actually the library part of "bluez" and "systemd-libs" is the library part of "systemd". I could not find a single package ending in "-libs" which doesn't also have some kind of "host package" above it.

On the other hand it seems like Arch doesn't do splitting of library and tools if not needed. Example:
https://archlinux.org/packages/community/x86_64/libaacs/
which ships whole library, include files and the binary tool "aacs_info".

So if that's common practice then a "netceiver-libs" package would practically also need some "netceiver" package which then would contain some kind of useful application to directly do something with the netceiver.

I'm also wondering if it is a good idea to keep "libmcli" as the name of the shared object file. Better make it "libnetceiver.so.*" then?

**Comment by pbiering** [comment]
RPM based distributions have a lot of "-libs", derivated from base package. I would still vote for reponame "netceiver-client", which depending the on distribution can either be used to create a all-in-one (tools and libs) binary package named "netceiver-client" or e.g. for RPM based split into "netceiver-client-libs" and "netceiver-client-utils".

And yes, the included lib should be named then "libnetceiver-client.so.*".

With this approach there would be also on conflict expected in case of "netceiver-server" will suddenly appear.

**Comment by M-Reimer** [comment]
I don't see that happen that we ever get "netceiver-server". This would require Reel to publish the sources and, as far as I know, they basically no longer exist. I actually wonder why their domain name is still reachable (with broken certificates but reachable). And even if this happened, the "server implementation" could have any kind of name. Maybe it's named "netceiverd" or "mclid". Who knows.

But this brings me to another point: I actually like the idea of having a new repo prefixed with "netceiver-" as this makes it easy to create more of those when needed. I'm currently thinking of "netceiver-firmwares". As far as I know the only source for them is a reelbox.org SVN and I'm unsure if this is still reachable at all. Probably a good idea to create a mirror for those. To keep existing hardware working, we for sure need a reliable source to get the firmware packages.

So back to topic. Let's have a prefix of "netceiver-". Would it be OK for you to name it "netceiver-libs" (as the primary content is the library which is needed for the actual "clients"). This would somewhat guarantee that the "primary package" probably would be named "netceiver-libs" through all distributions (I would not split this package for Arch just to separate the small utils). And if someone wants to have a "netceiver-utils", he can still split that off.

Otherwise we would have "netceiver-client-libs" and "netceiver-client-utils" for "RPM based" and just "netceiver-client" for Arch or other distributions that don't want to split the package.

**Comment by pbiering** [comment]
Idea behind of "netceiver-client" was to name the Source RPM equal to github repo name containing libs and tools as today in in the mcli source tree, anyhow not able be separated without strong dependency. Only the binary packages will have suffix "-libs" and "-utils". Having Source RPM named "netceiver-libs" then one could not expect that there are tools inside and that the library is client related only.

But finally the one creating the spin-off will decide...

**Comment by M-Reimer** [comment]
Just to simplify to decide on a good name I have searched a few that fall in to the category of "library + tools". Can you please check how one of these is handled for RPM creation?

https://archlinux.org/packages/community/x86_64/libaacs/
https://archlinux.org/packages/community/x86_64/libabigail/
https://archlinux.org/packages/extra/x86_64/libabw/
https://archlinux.org/packages/extra/x86_64/libaccounts-glib/

All these have some tools coming with them and Arch just decided to create one big package.

If you ask me, then my currently preferred name for the library is "libnetceiver". Simple. To the point. Easy to remember.

I did try to get some feedback from other users: https://www.vdr-portal.de/forum/index.php?thread/135589-library-von-plugin-absetzen-namensfrage/ But I guess not enough people are still interested in the netceiver hardware.

Do you think it could be a good idea to move the tools to their own repo or better keep tools and library in one repo? Probably it is not the best idea to fragment more than needed?

**Comment by pbiering** [comment]
pkgs.org is very useful for RPM lookup
- libaacs: base+devel+utils
  - utils contains: Test utilities for libaacs
- libabigail -> no sub-packages
  - some tools inside, but imho more internal ones
- libabw -> no sub-packages
  - no tools inside
- libaccounts-glib -> base+devel+doc
  - some tools inside, but imho more internal ones

so having a library package called "libnetceiver" can be ok, but the the tools which are not library-related but using library should then named somehow "netceiver-client-tools".

**Comment by M-Reimer** [comment]
> so having a library package called "libnetceiver" can be ok, but the the tools which are not library-related but using library should then named somehow "netceiver-client-tools".

Would this be the job of the packager or would you actually prefer the tools to be in their own repo? Would just "netceiver-tools" be OK, too?

**Comment by pbiering** [comment]
> > so having a library package called "libnetceiver" can be ok, but the the tools which are not library-related but using library should then named somehow "netceiver-client-tools".
> 
> Would this be the job of the packager or would you actually prefer the tools to be in their own repo? Would just "netceiver-tools" be OK, too?

Good question, initially one can create two different packages out of the same sources, later one can think of splitting the upstream repos....or still think about the proposal made earlier:
   
netceiver-libs, netceiver-tools and netceiver-devel, ongoing feeded from one repository.

**Comment by M-Reimer** [comment]
I had a deeper look into this topic and currently it seems to be at least challenging to correctly and nicely split this apart. There are plenty symlinks across the "mcast" directory. So I think I try with an "intermediate" step at first.

I've created https://github.com/vdr-projects/libnetceiver as a working repo. My goal would be:
- Move the "mcast" directory in the VDR plugin source to this new repo
- Drop "mcast/netcv2dvbip" and "mcast/dvbloop". The first one should be its own project and the second one probably no longer is relevant at all (missing kernel part for dvbloop). @pbiering Do you use either of the two?
- Add "pkg-config" file to allow other projects to find "libnetceiver" and get the required build configuration.
- Create a new Makefile to build both, library and tools and "make install" them.

I think having VDR plugin and "all the other stuff" separated would be a first step into the right direction. I would prepare the Makefile to make it easy to install (and so package) the tools separately as I would probably package them separately as "libnetceiver-tools". Making the tools dynamically link against "libnetceiver.so" could then be a future improvement step.

**Comment by M-Reimer** [comment]
Just as a reference. This one is an example where Arch splits a library package.
https://github.com/archlinux/svntogit-packages/blob/packages/libcamera/trunk/PKGBUILD

**Comment by M-Reimer** [comment]
Actually it was easier to link the tools dynamically against the library than I thought. There is still plenty of work needed to make this work as it should, but it is possible to create the library and compile the tools against it.

BUT: I found the first problem. https://github.com/vdr-projects/libnetceiver/commit/3872cb5febf396815d8ecf9fbe1c709bd5c6e5cd
I had to disable your "logmasks" for now as an "extern int" will not work well with libraries.

This either needs fixing or removing. One possible way for fixing could be to export functions to set the global variables and then use them from the VDR plugin to communicate the masks to the library.
